sw/host/opentitanlib/src/transport/errors.rs

```diff
@@ -0,0 +1,137 @@
+    NoDevice,
+    #[error("Found multiple USB devices, use --serial")]
+    MultipleDevices,
+    #[error("USB error: {0}")]
+    UsbGenericError(String),
+    #[error("Error opening USB device: {0}")]
+    UsbOpenError(String),
+    #[error("Transport does not support {0} instance {1}")]
+    InvalidInstance(TransportInterfaceType, String),
+    #[error("Encountered non-unicode device path")]
+    UnicodePathError,
+    #[error("Error opening {0}: {1}")]
+    OpenError(String, String),
+    #[error("Error reading from {0}: {1}")]
+    ReadError(String, String),
+    #[error("FPGA programming failed: {0}")]
+    FpgaProgramFailed(String),
+    #[error("Invalid pin strapping name \"{0}\"")]
+    InvalidStrappingName(String),
+    #[error("Transport does not support the requested operation")]
+    UnsupportedOperation,
+    #[error("Error communicating with FTDI: {0}")]
+    FtdiError(String),
+    #[error("Error communicating with debugger: {0}")]
+    CommunicationError(String),
+
+    // Include sub-enums for the various sub-traits of Tranport.
+    #[error("GPIO error: {0}")]
+    GpioError(crate::io::gpio::GpioError),
+    #[error("UART error: {0}")]
+    UartError(crate::io::uart::UartError),
+    #[error("SPI error: {0}")]
+    SpiError(crate::io::spi::SpiError),
+    #[error("SPI error: {0}")]
+    I2cError(crate::io::i2c::I2cError),
+}
+
+/// Enum value used by `TransportError::InvalidInstance`.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub enum TransportInterfaceType {
+    Gpio,
+    Uart,
+    Spi,
+    I2c,
+}
+
+impl std::fmt::Display for TransportInterfaceType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Gpio => write!(f, "GPIO"),
+            Self::Uart => write!(f, "UART"),
+            Self::Spi => write!(f, "SPI"),
+            Self::I2c => write!(f, "I2C"),
+        }
+    }
+}
+
+/// Return type to be used in `Transport` methods.
+pub type Result<T> = anyhow::Result<T, TransportError>;
+
+/// Convenience macro to be used in implementations of `Transport`.
+#[macro_export]
+macro_rules! bail {
+    ($err:expr $(,)?) => {
+        return Err($err.into());
+    };
+}
+
+/// Convenience macro to be used in implementations of `Transport`.
+#[macro_export]
+macro_rules! ensure {
+    ($cond:expr, $err:expr $(,)?) => {
+        if !$cond {
+            return Err($err.into());
+        }
+    };
+}
+
+/// Function for conveniently wrapping (the flat string representation of) an anyhow::Error in a
+/// TransportError.
+///
+/// Example usage:
+/// serialport::available_ports().wrap(UartError::EnumerationError)?;
+pub trait WrapInTransportError<T> {
+    fn wrap<S: Into<TransportError>>(self, kind: impl FnOnce(String) -> S) -> Result<T>;
+}
+
+impl<T, E: ToString> WrapInTransportError<T> for anyhow::Result<T, E> {
+    fn wrap<S: Into<TransportError>>(self, kind: impl FnOnce(String) -> S) -> Result<T> {
+        match self {
```

Comment:
This can be `map_err`.

Reply:
Yes it can.  I did not know of that method.